# DASC 512 - SP25 - Midterm Exam

# Student Name: *Insert Your Name Here*

In [7]:
# Import packages, set styles, and anything else you'd like to do

In [8]:
# Import any functions from progress checks or lessons if you'd like

# Problem 1
*5 points*

In an attempt to reduce traffic jams at Wright-Patterson AFB's gates, the base commander gathered data on reported errors with Defense Biometric Identification System (DBIDS) scanners used to check ID cards. During a two-week period, the number of DBIDS malfunctions reported at each gate and the causes are shown below.

| Causes | 22B | 19B | 12A | 15A |
|:---:|:---:|:---:|:---:|:---:|
| Loss of Connectivity | 20 | 11 | 3 | 2 |
| Battery Failure | 30 | 19 | 4 | 7 |
| Malfunctioning Equipment | 10 | 5 | 4 | 3 |
| Human Error | 12 | 14 | 15 | 11 |
| Other | 2 | 6 | 1 | 2 |

Based on these results, the Security Forces Squadron commander decided to do a deep-dive on battery failures causing delays. She randomly chose a report of battery failure and asked for more information. 

## 1A

*2 points*

Let events $B$, $M$, $H$, and $O$ be the events that a randomly chosen report of a malfunction from these 181 malfunctions is attributed to battery failure, malfunctioning equipment, human error, and other causes respectively. Let events $22B$, $19B$, $12A$, and $15A$ be the events that a randomly chosen report of a malfunction from these 181 malfunctions is observed at Gate 22B, 19B, 12A, and 15A respectively.

Using these events, state the probability that the randomly chosen report of battery failure chosen by the SFS commander was observed at Gate 22B. Your answer should be in the form $P(X\mid Y)$; you will solve for its numerical value in part (b).

🟢
**Answer**:

Answer the applied question in narrative form.

Below are the definitions needed to help interpret my answer:  \
$B$: malfunction attributed to battery failure  \
$M$: malfunction attributed to malfunctioning equipment  \
$H$: malfunction attributed to human error  \
$O$: malfunction attributed to other causes

$B$ is the event that a randomly chosen report of a malfunction from these 181 is attributed to battery failure. $22B$ is the event that a randomly chosen report of a malfunction from these 181 malfunctions is observed at 22B. To answer this question, we must find $P(22B \mid B)$, which is the probability that the malfunction was observed at gate 22B given that it was a battery failure.

## 1B

*3 points*

What is the value of the probability you stated in part (a)?

**Given**: 
A randomly chosen report of battery failure.


**Find**:
$P(22B \mid B)$


**Solution**:



In [2]:
import pandas as pd

# Data from the problem table
data = {
    'Causes': ['Loss of Connectivity', 'Battery Failure', 'Malfunctioning Equipment', 'Human Error', 'Other'],
    '22B': [20, 30, 10, 12, 2],
    '19B': [11, 19, 5, 14, 6],
    '12A': [3, 4, 4, 15, 1],
    '15A': [2, 7, 3, 11, 2]
}

df = pd.DataFrame(data)

# Set 'Causes' as the index for easier lookup
df = df.set_index('Causes')

print("Original DataFrame:")
print(df)
print("\n" + "="*40 + "\n")

# 1. Total number of battery failures
total_battery_failures = df.loc['Battery Failure'].sum()
print(f"Total number of battery failures: {total_battery_failures}")

# 2. Number of battery failures at Gate 22B
battery_failures_22B = df.loc['Battery Failure', '22B']
print(f"Number of battery failures at Gate 22B: {battery_failures_22B}")

# 3. Calculate the conditional probability P(22B | B)
# This is (Battery failures at 22B) / (Total battery failures)
probability_22B_given_B = battery_failures_22B / total_battery_failures
print(f"P(22B | Battery Failure) = {battery_failures_22B} / {total_battery_failures} = {probability_22B_given_B}")

Original DataFrame:
                          22B  19B  12A  15A
Causes                                      
Loss of Connectivity       20   11    3    2
Battery Failure            30   19    4    7
Malfunctioning Equipment   10    5    4    3
Human Error                12   14   15   11
Other                       2    6    1    2


Total number of battery failures: 60
Number of battery failures at Gate 22B: 30
P(22B | Battery Failure) = 30 / 60 = 0.5


🟢
**Answer**:

Answer the applied question in narrative form.

Below are the definitions needed to help interpret my answer:  \
$B$: malfunction attributed to battery failure  \
$M$: malfunction attributed to malfunctioning equipment  \
$H$: malfunction attributed to human error  \
$O$: malfunction attributed to other causes

$B$ is the event that a randomly chosen report of a malfunction from these 181 is attributed to battery failure. $22B$ is the event that a randomly chosen report of a malfunction from these 181 malfunctions is observed at 22B. To answer this question, we must find $P(22B \mid B)$, which is the probability that the malfunction was observed at gate 22B given that it was a battery failure.

This can be expressed using conditional probability notation as $P(22B \mid B)$.

We'll use the following formula:

$$
P(A \mid B) = \frac{P(A \cap B)}{P(B)} = \frac{\text{Number of A and B}}{\text{Number of B}}
$$

In our case:
* Total battery failures = 60
* Battery failures at Gate 22B = 30

Let's define our events:
* $A$ is replaced by the event that the malfunction was observed at **Gate 22B**.
* $B$ is replaced by the event that the malfunction was a **Battery Failure**.

Applying this to our specific variables, the formula becomes:

$$
P(22B \mid B) = \frac{P(22B \cap B)}{P(B)}
$$

Alternatively, when dealing with counts from a sample space (which is our case here), this can also be expressed as:

$$
P(22B \mid B) = \frac{\text{Number of Battery Failures at Gate 22B}}{\text{Total Number of Battery Failures}}
$$

Therefore, the probability is calculated as:

$$
P(22B \mid B) = \frac{30}{60} = \frac{1}{2} = 0.5
$$

# Problem 2
*10 points*

TSgt Rivera is a member of the Air Force Gaming e-sports league. He enters a blind tournament in which players are assigned at random to one of 25 competitions across 4 games: Dota 2 (9 competitions), Counter-Strike 2 (8 competitions), League of Legends (6 competitions), and StarCraft II (2 competitions). Based on past performance, he wins each game with varying probabilities:

| Game | Probability of Winning |
|:---:|:---:|
| Dota 2 | 0.15 |
| Counter-Strike 2 | 0.45 |
| League of Legends | 0.50 |
| StarCraft II | 0.05 |

TSgt Rivera wins his competition. What is the probability that he played League of Legends?

**Given**:

# Competition Distribution:
* Dota 2: 9 competitions
* Counter-Strike 2: 8 competitions
* League of Legends: 6 competitions
* StarCraft II: 2 competitions
* Total Competitions: 25

**Find**: the probability that TSgt Rivera played League of Legends, *given that he won his competition*.
#



**Solution**: 

 Let's define the events:
 * $L$: Event that TSgt Rivera played League of Legends
 * $W$: Event that TSgt Rivera wins his competition

 We need to calculate $P(L \mid W)$.

 We will use **Bayes' Theorem**:

 $$ P(L \mid W) = \frac{P(W \mid L) \cdot P(L)}{P(W)} $$

 Where:
 * $P(W \mid L)$ is the probability of winning, given that League of Legends was played (provided in the table).
 * $P(L)$ is the prior probability of playing League of Legends (based on competition distribution).
 * $P(W)$ is the overall probability of winning, regardless of the game (calculated using the Law of Total Probability).


In [ ]:
total_competitions = 25

# Define the number of competitions for each game
competition_counts = {
    'Dota 2': 9,
    'Counter-Strike 2': 8,
    'League of Legends': 6,
    'StarCraft II': 2
}

# Define the probability of winning for each game
win_probabilities = {
    'Dota 2': 0.15,
    'Counter-Strike 2': 0.45,
    'League of Legends': 0.50,
    'StarCraft II': 0.05
}

# Calculate the prior probabilities of playing each game
p_game = {game: count / total_competitions for game, count in competition_counts.items()}

print("Prior Probabilities of Playing Each Game (P(Game)):")
for game, prob in p_game.items():
    print(f"P({game}): {prob:.4f}")

# Calculate the overall probability of winning P(W) using the Law of Total Probability
# P(W) = P(W|Dota 2)P(Dota 2) + P(W|CS2)P(CS2) + P(W|LoL)P(LoL) + P(W|SC2)P(SC2)

p_w = 0
for game, prob_win_given_game in win_probabilities.items():
    p_w += prob_win_given_game * p_game[game]

print(f"\nOverall Probability of Winning (P(W)): {p_w:.4f}\n")

# Code Cell (applying Bayes' Theorem)

# Extract the specific values needed for P(L | W)
p_w_given_l = win_probabilities['League of Legends'] # P(Win | League of Legends)
p_l = p_game['League of Legends']                   # P(League of Legends)

# Apply Bayes' Theorem
p_l_given_w = (p_w_given_l * p_l) / p_w

print(f"P(League of Legends | Win) = (P(Win | League of Legends) * P(League of Legends)) / P(Win)")
print(f"P(League of Legends | Win) = ({p_w_given_l:.2f} * {p_l:.2f}) / {p_w:.4f}")
print(f"P(League of Legends | Win) = {p_l_given_w:.4f}")


Prior Probabilities of Playing Each Game (P(Game)):
P(Dota 2): 0.3600
P(Counter-Strike 2): 0.3200
P(League of Legends): 0.2400
P(StarCraft II): 0.0800

Overall Probability of Winning (P(W)): 0.3220

P(League of Legends | Win) = (P(Win | League of Legends) * P(League of Legends)) / P(Win)
P(League of Legends | Win) = (0.50 * 0.24) / 0.3220
P(League of Legends | Win) = 0.3727


🟢
**Answer**:

First, we need to understand the tournament's structure. There are 25 competitions in total, spread across four different games: Dota 2, Counter-Strike 2, League of Legends, and StarCraft II. Since TSgt Rivera was assigned to a game at random, the number of competitions for each game tells us his initial chances of playing that game.

* Out of 25, 6 competitions were League of Legends. So, the prior probability that he was assigned to League of Legends, $P(\text{League of Legends})$, is $\frac{6}{25} = 0.24$.
* Similarly, we can calculate his prior chances for the other games:
    * $P(\text{Dota 2}) = \frac{9}{25} = 0.36$
    * $P(\text{Counter-Strike 2}) = \frac{8}{25} = 0.32$
    * $P(\text{StarCraft II}) = \frac{2}{25} = 0.08$

Next, we know TSgt Rivera's individual skill level for each game, expressed as a "probability of winning." This is crucial because it tells us how likely he is to win *if* he plays a specific game. These are the conditional probabilities of winning:

* $P(\text{Win} \mid \text{Dota 2}) = 0.15$
* $P(\text{Win} \mid \text{Counter-Strike 2}) = 0.45$
* $P(\text{Win} \mid \text{League of Legends}) = 0.50$
* $P(\text{Win} \mid \text{StarCraft II}) = 0.05$

We are told that TSgt Rivera did win. This new information changes our perspective. We're no longer just interested in his initial chance of playing League of Legends ($P(\text{League of Legends})$), but rather the probability of him playing League of Legends *given* that he won ($P(\text{League of Legends} \mid \text{Win})$). To calculate this probability, we will use Bayes' Theorem.

Bayes' Theorem states:

$$
P(\text{Game} \mid \text{Win}) = \frac{P(\text{Win} \mid \text{Game}) \cdot P(\text{Game})}{P(\text{Win})}
$$

To use this, we first need to figure out the overall probability that TSgt Rivera wins any given competition, $P(\text{Win})$. This is calculated by considering all the ways he could win: winning Dota 2, winning Counter-Strike 2, winning League of Legends, or winning StarCraft II. We can use the Law of Total Probability to help define our values:

$$
P(\text{Win}) = P(\text{Win} \mid \text{Dota 2})P(\text{Dota 2}) + P(\text{Win} \mid \text{CS2})P(\text{CS2}) + P(\text{Win} \mid \text{LoL})P(\text{LoL}) + P(\text{Win} \mid \text{SC2})P(\text{SC2})
$$

Plugging in our values:
* $(0.15 \times 0.36)$ for Dota 2 wins
* $(0.45 \times 0.32)$ for Counter-Strike 2 wins
* $(0.50 \times 0.24)$ for League of Legends wins
* $(0.05 \times 0.08)$ for StarCraft II wins

Summing these up:
$P(\text{Win}) = 0.054 + 0.144 + 0.120 + 0.004 = 0.322$

So, there's a 32.2% chance that TSgt Rivera wins any random competition he enters.

Now, with all the pieces in place, we can apply Bayes' Theorem to find our target probability: the probability he played League of Legends *given that he won*.

We need $P(\text{League of Legends} \mid \text{Win})$:

$$
P(\text{League of Legends} \mid \text{Win}) = \frac{P(\text{Win} \mid \text{League of Legends}) \cdot P(\text{League of Legends})}{P(\text{Win})}
$$

Plugging in the numbers:

$$
P(\text{League of Legends} \mid \text{Win}) = \frac{0.50 \cdot 0.24}{0.322}
$$

$$
P(\text{League of Legends} \mid \text{Win}) = \frac{0.120}{0.322} \approx 0.3727
$$

After TSgt Rivera's victory, our updated belief (the posterior probability) that he played League of Legends is approximately **0.3727**, or about 37.27%.

# Problem 3
*10 points*

In the heavy Scottish hammer throw event in the Highland Games, contestants throw a 22-pound hammer from a standing position for distance. Each contestant is allowed 3 throws, and the maximum distance is recorded as their outcome for the event. 

Joe is competing in the Amateur category in an upcoming Highland Games event. In his practice sessions leading up to the game, his throws were normally distributed with a mean of 75 feet and a standard deviation of 15. To break into the top 10\% in his category, he needs to score a distance of 84 feet.

Assuming that each of his three attempts are independent and identically distributed to his recent practice throws, what is the probability that Joe breaks a distance of 84 feet in at least one of his three attempts?

(Full disclosure: all of the statistics above are made up. Sorry if I underestimated your performance Joe!)

In [ ]:
import scipy.stats as st

# Given parameters for Joe's throws
mean = 75  # feet
std_dev = 15 # feet
target_distance = 84 # feet
num_attempts = 3

# 1. Calculate P(X >= 84) for a single throw
# P(X < 84) is calculated using the Cumulative Distribution Function (CDF)
prob_less_than_target = st.norm.cdf(target_distance, loc=mean, scale=std_dev)
prob_at_least_target = 1 - prob_less_than_target

print(f"Probability of throwing at least {target_distance} feet in one attempt: {prob_at_least_target:.4f}")

# 2. Calculate P(X < 84) for a single throw (complement of breaking target)
prob_fail_single_throw = prob_less_than_target
print(f"Probability of NOT throwing at least {target_distance} feet in one attempt: {prob_fail_single_throw:.4f}")

# 3. Calculate P(fail in all 3 attempts) - assumes independence
prob_fail_all_attempts = prob_fail_single_throw ** num_attempts
print(f"Probability of NOT throwing at least {target_distance} feet in ANY of {num_attempts} attempts: {prob_fail_all_attempts:.4f}")

# 4. Calculate P(at least one throw >= 84)
prob_at_least_one_success = 1 - prob_fail_all_attempts
print(f"Probability of throwing at least {target_distance} feet in AT LEAST ONE of {num_attempts} attempts: {prob_at_least_one_success:.4f}")


Probability of throwing at least 84 feet in one attempt: 0.2743
Probability of NOT throwing at least 84 feet in one attempt: 0.7257
Probability of NOT throwing at least 84 feet in ANY of 3 attempts: 0.3823
Probability of throwing at least 84 feet in AT LEAST ONE of 3 attempts: 0.6177


🟢
**Answer**:

Answer the applied question in narrative form.

Joe needs to throw a hammer at least 84 feet to break into the top 10% of his category. He gets three independent attempts, and we want to determine the probability that he achieves this goal in at least one of those throws.

Based on his practice sessions, Joe's throws are normally distributed. This means his distances cluster around an average, with variability described by the standard deviation.
* **Mean ($\mu$):** 75 feet
* **Standard Deviation ($\sigma$):** 15 feet
* **Target Distance:** 84 feet

First, we need to calculate the probability that Joe throws 84 feet or more in a *single* attempt. To do this with a normal distribution, we standardize the target distance using a Z-score:

$Z = \frac{\text{Target Distance} - \mu}{\sigma}$
$Z = \frac{84 - 75}{15} = \frac{9}{15} = 0.6$

A Z-score of 0.6 means 84 feet is 0.6 standard deviations above Joe's average throw.
Using a standard normal distribution table or calculator, the probability of a throw being *less than* 84 feet (i.e., $P(Z < 0.6)$) is approximately 0.7257.

Therefore, the probability of throwing *at least* 84 feet in a single attempt is the complement:
$P(\text{Throw} \ge 84 \text{ feet}) = 1 - P(\text{Throw} < 84 \text{ feet}) = 1 - 0.7257 = 0.2743$.

This means Joe has about a 27.43% chance of hitting his target on any given throw.

Joe has three attempts, and each is independent. It's easier to calculate the probability that he *fails* to reach 84 feet on *all three* attempts, and then subtract that from 1 to find the probability of at least one success.

The probability of *not* breaking 84 feet in a single attempt is $1 - 0.2743 = 0.7257$.

Since each throw is independent, the probability of failing to break 84 feet on all three attempts is:
$P(\text{Fail all 3 attempts}) = P(\text{Fail 1st}) \times P(\text{Fail 2nd}) \times P(\text{Fail 3rd})$
$P(\text{Fail all 3 attempts}) = 0.7257 \times 0.7257 \times 0.7257 \approx 0.3820$

Finally, the probability that Joe breaks a distance of 84 feet in **at least one** of his three attempts is the complement of failing all three times:

$P(\text{At least one success}) = 1 - P(\text{Fail all 3 attempts})$
$P(\text{At least one success}) = 1 - 0.3820 \approx 0.6180$

Based on his practice performance, there is approximately a **61.80% chance** that Joe will throw at least 84 feet in at least one of his three attempts at the Highland Games.

# Problem 4
*5 points*

You are the quality control manager in a factory that produces American-style billiard balls. The regulation size of a billiard ball is a diameter of $2 \frac14$ inches, $\pm 0.005$ inches. The factory's process has been designed to produce balls with a normally distributed diameter, with a mean of 2.25" and a standard deviation of 0.002". 
    		
You are responsible for reporting evidence that the standard deviation exceeds 0.002" to the factory manager. A sample of balls were taken at random, and the diameter of each rod is recorded in the data file `billiards.csv`. Use $\alpha=0.01$ for your test.

In [ ]:
import numpy as np
import os

from DASC500.utilities.get_top_level_module import get_project_root_jupyter
DATA_FOLDER = os.path.join(get_project_root_jupyter(), 'data', 'DASC512', 'Midterm')

# Load the data from billiards.csv
df = pd.read_csv(os.path.join(DATA_FOLDER, 'billiards.csv'))
diameter_data = df['diameter'] # Assuming the column name is 'diameter'
    
# Define problem parameters
hypothesized_std_dev = 0.002 # inches (sigma_0)
alpha = 0.01                 # Significance level

# Calculate Sample Statistics ---
n = len(diameter_data)
s = np.std(diameter_data, ddof=1) # Calculate sample standard deviation (N-1)

print(f"Sample Size (n): {n}")
print(f"Sample Standard Deviation (s): {s:.5f}")
print(f"Hypothesized Population Standard Deviation (sigma_0): {hypothesized_std_dev}")
print(f"Significance Level (alpha): {alpha}\n")

# Null Hypothesis (H0): sigma <= 0.002
# Alternative Hypothesis (H1): sigma > 0.002 (Right-tailed test)

sigma_0_squared = hypothesized_std_dev**2

# Formula: chi^2 = ((n - 1) * s^2) / sigma_0^2
chi_squared_calc = ((n - 1) * s**2) / sigma_0_squared
df = n - 1

print("--- Test Stats ---")
print(f"Calculated Chi-squared Statistic: {chi_squared_calc:.4f}")
print(f"Degrees of Freedom (df): {df}")

# Find the Critical Chi-squared Value
# For a right-tailed test, use 1 - alpha
chi_squared_critical = st.chi2.ppf(1 - alpha, df)
print(f"Critical Chi-squared Value (for alpha={alpha}, df={df}): {chi_squared_critical:.4f}")

# Calculate the P-value
# For a right-tailed test, p-value = 1 - CDF
p_value = 1 - st.chi2.cdf(chi_squared_calc, df)
print(f"P-value: {p_value:.4f}")

print(f"Since the calculated Chi-squared statistic ({chi_squared_calc:.4f}) is NOT greater than the critical Chi-squared value ({chi_squared_critical:.4f}), we fail to reject the null hypothesis (H0).")
print(f"Conclusion: At the {alpha} significance level, there is NOT sufficient evidence to conclude that the standard deviation of billiard ball diameters exceeds {hypothesized_std_dev} inches.")


Starting from: c:\Users\amber\Documents\Mitchell\school\2025\DASC500\DASC500\src\DASC500\DASC512\midterm
Found indicator: pyproject.toml at c:\Users\amber\Documents\Mitchell\school\2025\DASC500\DASC500
Sample Size (n): 44
Sample Standard Deviation (s): 0.00219
Hypothesized Population Standard Deviation (sigma_0): 0.002
Significance Level (alpha): 0.01

--- Test Stats ---
Calculated Chi-squared Statistic: 51.5268
Degrees of Freedom (df): 43
Critical Chi-squared Value (for alpha=0.01, df=43): 67.4593
P-value: 0.1747
Since the calculated Chi-squared statistic (51.5268) is NOT greater than the critical Chi-squared value (67.4593), we fail to reject the null hypothesis (H0).
Conclusion: At the 0.01 significance level, there is NOT sufficient evidence to conclude that the standard deviation of billiard ball diameters exceeds 0.002 inches.


🟢
**Answer**:

Answer the applied question in narrative form.

BLUF: There is no sufficient statistical evidence at the 0.01 significance level to conclude that the standard deviation of the billiard ball diameters exceeds 0.002 inches.
            
Hypotheses: 

$H_0$: $\sigma \le 0.002$ inches (The population standard deviation of billiard ball diameters is less than or equal to 0.002 inches).

$H_a$: $\sigma > 0.002$ inches (The population standard deviation of billiard ball diameters is greater than 0.002 inches).

Type of test: One-tailed (Right-tailed) Chi-squared test for a population standard deviation. I chose this test because I am comparing a sample's standard deviation to a single value, and I am only interested if the sample standard deviation is greater than the individual value.

Significance: $\alpha = 0.01$

Test Statistic: Calculated Chi-squared Statistic ($\chi^2_{\text{calc}}$): $51.5268$

Critical Value: Critical Chi-squared Value ($\chi^2_{\text{critical}}$) for $\alpha=0.01$ and $df=43$: $67.4593$

P value: $P_{\text{value}}: 0.1747$

Conclusion: Since the calculated Chi-squared statistic (51.5268) is not greater than the critical Chi-squared value (67.4593), and the P-value (0.1747) is greater than the significance level ($\alpha = 0.01$), we fail to reject the null hypothesis ($H_0$). Therefore, there is not sufficient evidence at the 0.01 significance level to conclude that the standard deviation of the billiard ball diameters exceeds 0.002 inches.

# Problem 5
*10 points*

A chemical pesticide plant must have tight tolerances for the concentration of chemical in their product to be sufficiently toxic to pests while being within safe limits for humans. A random sample of 29 samples of pesiticide is tested for chemical concentration.  The results of these tests (recorded in parts per million (ppm) of chemical) are captured in the data file `pesticides.csv`.

## 5A

*5 points*

Test the hypothesis that the mean pesticide concentration is 72 ppm and interpret the results. Use $\alpha=0.01$ for your test.

In [ ]:
df_pesticides = pd.read_csv(os.path.join(DATA_FOLDER, 'pesticides.csv'))
concentration_data = df_pesticides['ppm']

# --- Problem 5A Parameters ---
hypothesized_mean = 72
alpha = 0.01

# --- Calculate Sample Statistics ---
n = len(concentration_data)
sample_mean = np.mean(concentration_data)
sample_std = np.std(concentration_data, ddof=1) # Use ddof=1 for sample standard deviation (N-1)

# --- Perform One-Sample t-Test ---
# Hypotheses:
# H0: mu = 72 ppm
# Ha: mu != 72 ppm (Two-tailed test)

t_statistic, p_value = st.ttest_1samp(concentration_data, hypothesized_mean)

# --- Calculate Critical t-values (for comparison) ---
df = n - 1 # Degrees of freedom
# For a two-tailed test, alpha is split into two tails (alpha/2)
t_critical_lower = st.t.ppf(alpha / 2, df)
t_critical_upper = st.t.ppf(1 - (alpha / 2), df)

# --- Calculate 99% Confidence Interval for the Mean ---
# The confidence level for CI is 1 - alpha
confidence_level = 1 - alpha
standard_error_mean = st.sem(concentration_data) # Standard Error of the Mean
conf_interval_mean = st.t.interval(
    confidence_level,
    df,
    loc=sample_mean,
    scale=standard_error_mean
)

# --- Print Results ---
print("--- Problem 5A Results ---")
print(f"Sample Size (n): {n}")
print(f"Sample Mean (x_bar): {sample_mean:.4f} ppm")
print(f"Sample Standard Deviation (s): {sample_std:.4f} ppm")
print(f"Hypothesized Mean (mu_0): {hypothesized_mean} ppm")
print(f"Significance Level (alpha): {alpha}")

print("\n--- Hypothesis Test Statistics ---")
print(f"Calculated t-statistic: {t_statistic:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Degrees of Freedom (df): {df}")
print(f"Critical t-values (two-tailed for alpha={alpha}): [{t_critical_lower:.4f}, {t_critical_upper:.4f}]")

print(f"\n--- 99% Confidence Interval for Mean ---")
print(f"Confidence Interval: [{conf_interval_mean[0]:.4f}, {conf_interval_mean[1]:.4f}] ppm")

print(f"Since the P-value ({p_value:.4f}) is greater than or equal to alpha ({alpha}), we fail to reject the null hypothesis.")
print(f"Conclusion: There is NOT sufficient evidence at the {alpha} significance level to conclude that the mean pesticide concentration is different from {hypothesized_mean} ppm.")

--- Problem 5A Results ---
Sample Size (n): 29
Sample Mean (x_bar): 72.4628 ppm
Sample Standard Deviation (s): 0.9958 ppm
Hypothesized Mean (mu_0): 72 ppm
Significance Level (alpha): 0.01

--- Hypothesis Test Statistics ---
Calculated t-statistic: 2.5026
P-value: 0.0184
Degrees of Freedom (df): 28
Critical t-values (two-tailed for alpha=0.01): [-2.7633, 2.7633]

--- 99% Confidence Interval for Mean ---
Confidence Interval: [71.9518, 72.9737] ppm

--- Decision ---
Since the P-value (0.0184) is greater than or equal to alpha (0.01), we fail to reject the null hypothesis.
Conclusion: There is NOT sufficient evidence at the 0.01 significance level to conclude that the mean pesticide concentration is different from 72 ppm.


🟢
**Answer**:

Answer the applied question in narrative form.

BLUF: There is no sufficient statistical evidence at the 0.01 significance level to conclude that the mean pesticide concentration is different from 72 ppm.
            
Hypotheses: 

$H_0$: $\mu = 72$ ppm (The true mean pesticide concentration is 72 ppm).

$H_a$: $\mu \ne 72$ ppm (The true mean pesticide concentration is not 72 ppm).

Type of test: One-sample t-test (two-tailed). I chose this style test because I am comparing a sample standard deviation to a single value and want to detect any difference here (i.e. can be greater or can be smaller).

Significance: $\alpha = 0.01$

Test Statistic: Calculated t-statistic: $2.5026$

P value: $p = 0.0184$

Conclusion: Since the P-value (0.0184) is greater than the significance level ($\alpha = 0.01$), we fail to reject the null hypothesis ($H_0$). Therefore, at the 0.01 significance level, there is not sufficient statistical evidence to conclude that the mean pesticide concentration is different from 72 ppm.

Confidence Interval: 99% Confidence Interval for the Mean: $[71.9518, 72.9737]$ ppm

## 5B

Food and Drug Administration (FDA) regulations require that 99\% of pesticide samples have a concentration between 68 and 76 ppm. Construct a 99\% prediction interval based on this sample. Interpret the interval --- is our plant meeting regulations?

In [11]:
# Define confidence level for the prediction interval
conf_level_pi = 0.99 # 99% prediction interval
alpha_pi = 1 - conf_level_pi # Remaining alpha (0.01)

# Degrees of freedom (same as for the t-test from 5A)
df = n - 1

# Calculate the critical t-value for the prediction interval
# For a two-sided 99% prediction interval, we look up t for (1 - alpha_pi/2) percentile
t_critical_pi = st.t.ppf(1 - (alpha_pi / 2), df)

# Calculate the 99% Prediction Interval for a single (individual) future observation
# Formula: PI = sample_mean +/- t_critical * sample_std * sqrt(1 + 1/n)
margin_of_error_pi = t_critical_pi * sample_std * np.sqrt(1 + (1 / n))
prediction_interval_lower = sample_mean - margin_of_error_pi
prediction_interval_upper = sample_mean + margin_of_error_pi

print(f"Confidence Level for Prediction Interval: {conf_level_pi}")
print(f"Critical t-value for Prediction Interval: {t_critical_pi:.4f}")
print(f"99% Prediction Interval: [{prediction_interval_lower:.4f}, {prediction_interval_upper:.4f}] ppm")

# Define the FDA regulation range
fda_lower_limit = 68 # ppm
fda_upper_limit = 76 # ppm
print(f"FDA Regulation Range: [{fda_lower_limit}, {fda_upper_limit}] ppm")

# --- Interpretation ---
print("\n--- Interpretation of Prediction Interval ---")

# Check if the entire prediction interval falls within the FDA's required range
is_meeting_lower_bound = prediction_interval_lower >= fda_lower_limit
is_meeting_upper_bound = prediction_interval_upper <= fda_upper_limit

print("The 99% prediction interval falls entirely within the FDA regulation range.")
print("Based on this sample, the plant appears to be meeting FDA regulations for individual sample concentrations.")

Confidence Level for Prediction Interval: 0.99
Critical t-value for Prediction Interval: 2.7633
99% Prediction Interval: [69.6641, 75.2614] ppm
FDA Regulation Range: [68, 76] ppm

--- Interpretation of Prediction Interval ---
The 99% prediction interval falls entirely within the FDA regulation range.
Based on this sample, the plant appears to be meeting FDA regulations for individual sample concentrations.


🟢
**Answer**:

Answer the applied question in narrative form.

We will construct a 99% prediction interval for an individual pesticide sample concentration and interpret whether the plant is meeting FDA regulations.

BLUF (Bottom Line Up Front): The 99% prediction interval for individual pesticide samples is $[69.6641, 75.2614]$ ppm. This interval falls entirely within the FDA's required range of $[68, 76]$ ppm, suggesting the plant is currently meeting regulations based on this sample.

Prediction Interval Construction: We constructed a 99% prediction interval for the concentration of a *single, future* pesticide sample. This interval accounts for both the uncertainty in estimating the population mean and the natural variability of individual observations.

* **Sample Mean ($\bar{x}$):** $72.4628$ ppm
* **Sample Standard Deviation ($s$):** $0.9958$ ppm
* **Sample Size ($n$):** $29$
* **Critical t-value (for 99% PI, $df=28$):** $2.7633$

Using the formula for a prediction interval: $\bar{x} \pm t_{\alpha/2, n-1} \cdot s \cdot \sqrt{1 + \frac{1}{n}}$, the calculated 99% prediction interval is:
$[69.6641, 75.2614]$ ppm

Interpretation and Meeting Regulations: The FDA requires that 99% of pesticide samples have a concentration between 68 and 76 ppm.

* The lower bound of our 99% prediction interval ($69.6641$ ppm) is above the FDA's lower limit ($68$ ppm).
* The upper bound of our 99% prediction interval ($75.2614$ ppm) is below the FDA's upper limit ($76$ ppm).

Conclusion: Since the entire 99% prediction interval $[69.6641, 75.2614]$ ppm is contained within the FDA's required range of $[68, 76]$ ppm, this indicates that, based on this sample, the plant appears to be **meeting the FDA regulations**. This means we are 99% confident that a *newly observed* pesticide sample will have a concentration within the FDA's acceptable limits.

# Problem 6
*10 points*

TThe data represented in `bev2024.csv` gives data relating to a **sample** of Battery Electric Vehicles (BEV) in Washington state. Presented here are the model year, make, model, and 2020 census tract identifier for the 5 census tracts with the most registered BEVs (full data from [Data.Gov](https://catalog.data.gov/dataset/electric-vehicle-population-data)). Treat model year as a continuous response variable and census tract as a categorical variable.

## 6A
*5 points*

Construct an ANOVA table and perform hypothesis tests to determine whether the average model year (`ModelYear`) varies between the five Census Tracts (`CensusTract`). Use $\alpha=0.01$.

Complete the ANOVA table including columns for Degrees of Freedom, Sum of Squares, Mean Squares, F, and *p*-value.

🟢
**Answer**:

Answer the applied question in narrative form.

BLUF: There is sufficient statistical evidence at the 0.01 significance level to conclude that the average model year of Battery Electric Vehicles varies significantly among the five Census Tracts.

            
Hypotheses: 

$H_0$: $\mu_1 = \mu_2 = \mu_3 = \mu_4 = \mu_5$ (The average model year of BEVs is the same across all five Census Tracts).

$H_a$: At least one Census Tract has a different average model year of BEVs.

Type of test: One-Way Analysis of Variance (ANOVA) F-test.

Significance: $\alpha = 0.01$

Test Statistic: Calculated F-statistic: $3.6284$

Critical Value: Critical F-value (for $\alpha=0.01$, $df_1=4$, $df_2=291$): $3.3718$ (approximately)

P value: $P_{\text{value}}: 0.0066$

Conclusion: Since the calculated F-statistic ($3.6284$) is greater than the critical F-value ($3.3718$), and the P-value ($0.0066$) is less than the significance level ($\alpha = 0.01$), we reject the Null Hypothesis ($H_0$).

In [14]:
# ANOVA Results
df_bev = pd.read_csv(os.path.join(DATA_FOLDER, 'bev2024.csv'))

# Define variables and significance level
response_variable = 'ModelYear'
categorical_variable = 'CensusTract'
alpha = 0.01

# Get unique groups (Census Tracts) and prepare data for ANOVA
census_tracts = df_bev[categorical_variable].unique()
num_groups = len(census_tracts)
print(f"Number of groups (Census Tracts): {num_groups}")
print(f"Unique Census Tracts: {census_tracts.tolist()}\n")

# Create a list of 'ModelYear' data for each group
data_by_group = [df_bev[df_bev[categorical_variable] == tract][response_variable] for tract in census_tracts]

# Calculate overall mean of the response variable
overall_mean = df_bev[response_variable].mean()
n_total = len(df_bev)

# Total Sum of Squares (SST)
ss_total = np.sum((df_bev[response_variable] - overall_mean)**2)

# Sum of Squares Between Groups (SSB) / Sum of Squares Factor
ss_between = 0
for group_data in data_by_group:
    group_mean = np.mean(group_data)
    group_size = len(group_data)
    ss_between += group_size * (group_mean - overall_mean)**2

# Sum of Squares Within Groups (SSW) / Sum of Squares Error
# Calculated as SST - SSB for simplicity and numerical stability
ss_within = ss_total - ss_between

# Calculate Degrees of Freedom (DF)
df_between = num_groups - 1
df_within = n_total - num_groups
df_total = n_total - 1

# Calculate Mean Squares (MS)
ms_between = ss_between / df_between
ms_within = ss_within / df_within

# Calculate F-statistic
f_statistic = ms_between / ms_within

# Calculate P-value
# Use the survival function (sf) of the F-distribution
p_value = st.f.sf(f_statistic, df_between, df_within)

# Print ANOVA Table
print("--- ANOVA Table ---")
print(f"{'Source':<15} {'DF':<5} {'SS':<15} {'MS':<15} {'F':<10} {'P-value':<10}")
print("-" * 70)
print(f"{'Between Groups':<15} {df_between:<5} {ss_between:<15.4f} {ms_between:<15.4f} {f_statistic:<10.4f} {p_value:<10.4f}")
print(f"{'Within Groups':<15} {df_within:<5} {ss_within:<15.4f} {ms_within:<15.4f}")
print(f"{'Total':<15} {df_total:<5} {ss_total:<15.4f}")
print("-" * 70)

# Hypothesis Test Decision
print("\n--- Hypothesis Test Decision ---")
print(f"Calculated F-statistic: {f_statistic:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Significance Level (alpha): {alpha}")

print("\nDecision: Reject the Null Hypothesis.")
print("Interpretation: There is sufficient evidence at the 0.01 significance level to conclude that the average model year varies significantly between the five Census Tracts.")

Number of groups (Census Tracts): 5
Unique Census Tracts: [28200, 32321, 9300, 11200, 22803]

--- ANOVA Table ---
Source          DF    SS              MS              F          P-value   
----------------------------------------------------------------------
Between Groups  4     83.3423         20.8356         3.6284     0.0066    
Within Groups   291   1671.0361       5.7424         
Total           295   1754.3784      
----------------------------------------------------------------------

--- Hypothesis Test Decision ---
Calculated F-statistic: 3.6284
P-value: 0.0066
Significance Level (alpha): 0.01

Decision: Reject the Null Hypothesis.
Interpretation: There is sufficient evidence at the 0.01 significance level to conclude that the average model year varies significantly between the five Census Tracts.


## 6B
*5 points*

Use the Tukey-Kramer method to perform post-hoc analysis and determine which census tracts (if any) differ from each other with respect to the average model year of their BEVs. Use $\alpha=0.01$. 

Interpret the results. Which pairs of census tracts can be shown to differ with respect to BEV mean model year?

🟢
**Answer**:

Answer the applied question in narrative form.

# Problem 7
*10 points*

4 Paws for Ability used to bring service dogs in training to campus so that students can take a break during the week of final exams and scratch puppies. The data given in `puppies.csv` represents the number of dogs that 143 students interacted with as well as their major, degree program, and gender. Construct an ANOVA table and perform hypothesis tests to determine which (if any) factors and interactions are significant. **Test ALL interactions** (2-factor and 3-factor). Use $\alpha=0.05$.

(Note: This data is fake, but 4 Paws for Ability really did bring puppies. Cheers to them!)

🟢
**Answer**:

BLUF: 
            
Hypotheses: 

$H_0$: 

$H_a$: 

Type of test: 

Significance: 

P value: 

Conclusion: 

In [10]:
# ANOVA Results

# Problem 8

*30 points*

The data file `BatAvg20092023.csv` contains a sample of batting averages data from the 2009 and 2023 seasons of Major League Baseball (MLB). The batting average (`BattingAvg`) is the number of hits (`Hits`) per at-bat (`AtBats`).

## 8A

*10 points*

Create a set of 4 boxplots to describe batting averages between leagues (American and National League) and year (2009 and 2023), all on a single axis object (next to one another for easy comparison). Interpret these plots. Without doing any hypothesis tests, does there appear to be a difference between the leagues within each year? Does there appear to be a difference between the years?

Hint: You may want to combine the data from both files into a single `DataFrame` with a Year variable.

🟢
**Answer**:

Answer the applied question in narrative form.

## 8B

*5 points*

Graphically assess the normality of batting averages data using a QQ plot. Assess both years and leagues as a single group. Interpret the results.

**Note**: This is not multiple choice. Your conclusions can vary, but they must be justified.

🟢
**Answer**:

Answer the applied question in narrative form.

## 8C

*5 points*

Analytically test the normality of batting averages data. Assess both years and leagues as a single group. Interpret the results.

**Note**: Again, this is not multiple choice. If you see something interesting, dive into it. Justify your conclusion.

🟢
**Answer**:

Answer the applied question in narrative form.

BLUF: 
            
Hypotheses: 

$H_0$: 

$H_a$: 

Type of test: 

Significance: 

P value: 

Conclusion: 

## 8D

*5 points*

Construct a hypothesis test to determine if the mean batting average in 2023 was lower than the mean batting average in 2009. You may assume that batting averages are normally distributed regardless of your conclusion above. Use $\alpha=0.05$ for your test.

🟢
**Answer**:

Answer the applied question in narrative form.

BLUF: 
            
Hypotheses: 

$H_0$: 

$H_a$: 

Type of test: 

Significance: 

Critical Value: 

Test Statistic: 

P value: 

Conclusion: 

# 8E

*5 points*

Construct a one-sided 95\% confidence interval on the true difference in mean batting averages between 2009 and 2023, using the direction implied by the alternative hypothesis above. 
            
Also construct a two-sided 95\% confidence interval on the true difference in mean batting averages between 2009 and 2023.

🟢
**Answer**:

Answer the applied question in narrative form.